### ISSUE TITLE
plotting point cloud before FPS

### ISSUE THREAD

**Issue by gilad011** [GeeeG/CloudPose#6]
HI @GeeeG, 
I am running object_6d_pose.py file.
the resulted point clouds are: 
- green points from the object CAD model (am i right?) 
- image colored points for the aligned segmented point cloud after FPS.
couple questions:
1. I saw that the image colored cloud consist 256 points instead of 1024 that resulted from FPS, why is that?
2. why there are outlier points in the resulted cloud? the segmented areas looks good.
3. how can i visualize the cloud before FPS and after? I have trouble getting to it from the defined dataset object.

appreciate your help!
![Capture](https://user-images.githubusercontent.com/40541255/89508888-38da6a00-d7d7-11ea-8a7a-62fad87aae0c.PNG)

**Comment by GeeeG** [comment]
Hi @gilad011 ,
- Yes, the green points are from 3D object model
- The colored points is the input point cloud segment to the network, it contains 256 point which is obtained with FPS, and it is transformed with the 6D pose estimations.

1. The created training datasets in tfrecord format contains segment with 1024 points. However, for training the network, we only use 256 points out of 1024 for faster training, hence for testing (object_6d_pose.py), we test with 256 points. It is controlled by --num_point in object_6d_pose.py.
2. The outlier points come from the not perfect object segmentation in the test set.
3. For visualization before and after FPS, if you go to "segment_sample_FPS" in the "data_process_tools.py"
(https://github.com/GeeeG/CloudPose/blob/dd99d40ea01e379daafef22fca08419ed555b143/data_process_tools.py#L109)
you can see where FPS is called, you can add another output for before, e.g.
y_out["xyz_org"] = tf.boolean_mask(x['xyz'], label_mask_r)
Then you can access this segment and visualize it.

**Comment by gilad011** [comment]
@GeeeG thank you for the detailed comment!
1. is it means that I can not change it from 256 in testing if the network was trained with 256?
3. I actually already manage to do it the way you suggested:
![Capture_preFPS](https://user-images.githubusercontent.com/40541255/90113575-3c7f6b00-dd5a-11ea-8859-1e074eba5d29.PNG)

**Comment by GeeeG** [comment]
@gilad011 no worries!
Yes, if the network is trained with 256 points, it expects a 256 points input.
I tried to train the same network with 512, the performance is more or less the same, and it takes much longer to train.
I think what is important here is the surface geometry captured by point cloud, and 256 seems to be sufficient

**Comment by gilad011** [comment]
Thanks!
I saw that during training there is no evaluation procedure on validation-set at each epoch.
Is it something that you are going to add?

**Comment by GeeeG** [comment]
You can add the validation during training with something this (I removed it before uploading):
DATASET_SIZE = number of samples in your training set
train_size = int(0.9 * DATASET_SIZE)
val_size = int(0.1 * DATASET_SIZE)

full_dataset = tf.data.TFRecordDataset(train_filenames).shuffle(1000000)

tr_dataset = full_dataset.take(train_size)
tr_dataset = get_small_data(tr_dataset, batch_size=BATCH_SIZE, total_num_point=TOTAL_NUM_POINT)
tr_iterator = tr_dataset.make_initializable_iterator()

val_dataset = full_dataset.skip(train_size)
val_dataset = get_small_data(val_dataset, batch_size=BATCH_SIZE, total_num_point=TOTAL_NUM_POINT)
val_iterator = val_dataset.make_initializable_iterator()

Then you can add an other function name e.g. "eval_graph" similar to "train_graph" with the validation set, and call it after each epoch.
